doc/adr/0002-GraphQL.md

```diff
@@ -0,0 +1,317 @@
+  1. ***Paging*** supported by GraphQL clients; [pagination][pag] ([practices][relay]).
+  1. Also depending on how our GraphQL resolver functions would perform.
+
+#### Open-source libraries
+
+Language support for server-side and client-side implementation.
+
+* [Pluralsight course][course] mentions GraphQL being [language agnostic][code], [overall][impl].
+* Can we quickly confirm this for both clients and server sides?
+  * [Python][py] and TypeScript ([JavaScript][js]) for clients;
+  * [Java][java] for (trace-)server TSP reference implementation, our currently REST API.
+* [Pluralsight][course]: GraphQL open-sourced, with ***large*** [community][comm] involvement.
+
+#### Versioning
+
+According to the [Pluralsight course][course],
+
+* there is ***no API versioning required***,
+* because of queried fields flexibility;
+* non-breaking API changes enabling.
+* [GraphQL versioning][vers].
+
+#### Centralized server
+
+Possible for `trace-server` servicing?
+
+From the [Pluralsight course][course]:
+
+* GraphQL Server able to fetch data from ***legacy system***(s).
+  1. By means of writing per-field Resolver functions.
+  1. Batched (optimized) resolving supported; [batching][batch].
+  1. API gets itself documented through the developing GraphQL schema.
+
+## Decision
+
+### Proposed change
+
+Based on [Mermaid][mermaid] used [in VS Code][ext].
+
+#### Current architecture
+
+```mermaid
+graph LR
+  REST/http --> trace-server
+  trace-server-protocol --> REST/http
+  tsp-python-client --> trace-server-protocol
+  tsp-typescript-client --> trace-server-protocol
+  subgraph Clients
+    tsp-python-client
+    tsp-typescript-client
+  end
+  subgraph APIs
+    trace-server-protocol
+    REST/http
+  end
+  subgraph Servers
+    trace-server
+  end
+```
+
+Component names above are based on their repository names or directory:
+
+  1. [trace-server][inc]; reference implementation only, could be more servers.
+  1. [trace-server-protocol][tsp]; currently based on REST (over HTTP), may evolve.
+  1. [tsp-python-client][client-py]; TSP implementation to be used in clients written in Python.
+  1. [tsp-typescript-client][client-ts]; TSP implementation used in clients written in TypeScript.
+
+#### GraphQL for TSP API (over REST)
+
+```mermaid
+graph TD
+  REST/http --> trace-server
+  GraphQL --> GraphQL-Server
+  GraphQL-Server --> REST/http
+  trace-server-protocol --> REST/http
+  trace-server-protocol --> GraphQL
+  tsp-python-client --> trace-server-protocol
+  tsp-typescript-client --> trace-server-protocol
+  subgraph Clients
+    tsp-python-client
+    tsp-typescript-client
+  end
+  subgraph APIs
+    trace-server-protocol
+    REST/http
+    GraphQL
+  end
+  subgraph Servers
+    trace-server
+    GraphQL-Server
+  end
+```
+
+1. GraphQL would be offered as a TSP option alongside current REST endpoints.
+1. Resolver functions in GraphQL Server would rely on current REST endpoints.
```

Comment:
Is it required that the resolver translate the GraphQL queries to REST API calls that are currently implemented in the trace server? If I don't misunderstand it, that would add 1 serialization and 1 deserialization to the query execution path, which will increase the overall round-trip for a query.

Reply:
Yes only for the time being, and totally right; indeed. As previously shared, we will still be able to reconsider that approach eventually or at will. -Just not yet or for this initial ADR version of ours. (Or so was the proposal so far.)

Reply:
Thanks for clarifying.